refactor(app): add load labware, deck setup, tip pick up, point checks to cal check

b-cooper opened the pull request:
## overview

Add views and supporting business logic to back the robot calibration flow. Additions include:
guiding user through loading the correct tipracks on the deck, jogging and picking up a tip for the
active pipette, checking the three xy calibration points, and checking the z calibration.

This PR grew to encapsulate a fair amount of new code because of some extenuating circumstances. Apologies for the large scope, I'll do my best to summarize the changes below.

Closes #5033, Closes #5034, Closes #5036, Closes #5037

## changelog
- add robot calibration check components: `DeckSetup`, `TipPickUp`, `CheckXYPoint`, and `CheckHeight`
- add GET existing session behavior instead instead of creating a new robot cal check session by default every time. Flow is now if no existing session, try and create one. 409 conflicting for the POST should rarely if ever be hit.
- add POST update session behavior to handle all actions a user can dispatch from the newly implemented views that would update the robot cal check session. This includes: jogging, picking up tip, confirming tip is attached, claiming no tip is attached, checking all X/Y points in the flow, and checking the height at slot 5. 
- allow `mapActionToRequest` to return null which will be filtered out by the `mapToRobotApiRequest` implementation. This allows epics that use this operator to receive many action types and branch logic accordingly (some branches passing through unhandled to be filtered out)
- add some props to the `JogControls` component so that it can be reused in this flow and elsewhere. Both the step options and the mutable axes are parameterized now.

## review requests

- [ ] With feature flag on, start a robot calibration check session for a robot running edge
- [ ] it prompts you to load the correct tip racks onto your deck, given your attached pipettes
- [ ] the gantry then moves to the above A1 of the tiprack for your left pipette 
- [ ] you can then jog in all directions and step sizes to get the nozzle above the tip
- [ ] pick up the tip and confirm tip was attached
- [ ] proceed through checks at crosses on slots 1, 3 and, 7
- [ ] proceed to check deck height  at slot 5 

## risk assessment

Large footprint, but nearly all is behind a feature flag.  

There are two exceptions that do touch code that was already in place.

- the ability of the robot-api `mapActionToRequest` function to return null and get immediately filtered out. This is useful in allowing the `updateRobotCalibrationCheckSessionEpic` to handle many types input actions. This shouldn't have any effect at all to existing code, but technically touches any requests through the new robot-api rxjs operators machinery.

- the `JogControls` component that was being used in a couple places in the app, is now capable of supporting some axis and step size customization.

b-cooper committed:
add place tipracks step and stubbing mechanism

b-cooper committed:
refactor(app): add tiprack setup prompt to robot calibration check flow

Prompt user to load the required tipracks into the correct slots of the deck before proceeding to
check the robot's calibration

Closes #5033

b-cooper committed:
remove unnecessary spacing diff

b-cooper committed:
little lint

b-cooper committed:
unstyled full flow

b-cooper committed:
revert stubs

b-cooper committed:
build up update session

b-cooper committed:
add to cal action type

b-cooper committed:
test fetch epic

b-cooper committed:
update epic test with first action

b-cooper committed:
handle update with params passed

b-cooper committed:
add shim current step

b-cooper committed:
fix up tests and add temporary step shimmer

b-cooper committed:
wire up shim a bit more

b-cooper committed:
prepare pipette

b-cooper committed:
bit of lint

b-cooper committed:
wire jog up

b-cooper committed:
jog sending correct vector values

b-cooper committed:
add img assets

b-cooper committed:
assets and styling for jog to tiprack

b-cooper committed:
wire up inspecting tip

b-cooper committed:
add domain layer for confirm and invalidate tip

b-cooper committed:
style confirm tip attached propmt

b-cooper committed:
check x and y with parameterized jog controls

b-cooper committed:
imprt video assets

b-cooper committed:
ensure jog label logic

b-cooper committed:
move to point webm in place

b-cooper committed:
connect confirm step

b-cooper committed:
make sure video updates

b-cooper committed:
add check height

b-cooper committed:
placeholder spinner

b-cooper committed:
remove temporary shim for state update

b-cooper committed:
hide controls on video demos

b-cooper committed:
update all video assets and replace gifs with webm

b-cooper committed:
clean up lint

b-cooper committed:
button up flow errors

b-cooper committed:
bit of lint

b-cooper committed:
tests for tip pick up component

b-cooper committed:
test Check xy point component

b-cooper committed:
test check height component

b-cooper committed:
add tests for fetch and update actions

b-cooper committed:
test all actions

b-cooper committed:
update deck setup test

b-cooper committed:
update index component test

b-cooper committed:
add provider to connected component test

b-cooper committed:
test all session update epic actions

b-cooper committed:
refactor(app): add load labware, deck setup, tip pick up, point checks to cal check

Add views and supporting business logic to back the robot calibration flow. Additions include:
guiding user through loading the correct tipracks on the deck, jogging and picking up a tip for the
active pipette, checking the three xy calibration points, and checking the z calibration.

Closes #5033, Closes #5034, Closes #5036, Closes #5037

b-cooper committed:
remove a little lint

b-cooper committed:
delint css

b-cooper committed:
capital U

sfoster1 reviewed (changes requested):
Some changes from testing:

- In the deck view, the long labware names mean that you almost can’t see the relevant part (the tip rack volume); in addition, resizing the window resizes the text rather than displaying more of it. Could we have a double row of text there?
- You can still queue up jogs by pressing really fast
- If you fail to pick up tip and press try again the pipette doesn’t move back to the tip rack
- All-capsed tip rack names display “10 ML” which I assume is the capital mu again

mcous commented on app/src/components/CheckCalibration/DeckSetup.js:
```diff
@@ -0,0 +1,121 @@
+  )
+}
+
+type TiprackRenderProps = { labwareDef: LabwareDefinition2, slotDef: DeckSlot }
```
Can this be an exact object?

mcous commented on app/src/components/CheckCalibration/TipPickUp.js:
```diff
@@ -0,0 +1,153 @@
+  )
+  const dispatch = useDispatch<Dispatch>()
+
+  function jog(axis: JogAxis, direction: JogDirection, step: JogStep) {
```
[No change requested] A lot of these action dispatchers look common. At some point a `useCalibrationActions` (or whatever name) hook that does the `robotName` + `pipetteId` + `dispatch` binding could be cool and reduce the testing burden a little

b-cooper replied:
That's a good idea, and it plugs into some of the simplifications of the http interface that might consolidate the update epic logic as well.

mcous commented on app/src/components/CheckCalibration/__tests__/DeckSetup.test.js:
```diff
@@ -0,0 +1,73 @@
+
+  beforeEach(() => {
+    // NOTE: deliberately not testing RobotWorkSpace internals here
+    MockRobotWorkSpace.mockReturnValue(null)
```
I've realized I can shortcut these kinds of component mocks at the mock call itself to reduce test setup boilerplate. If it works here, it would look like:

```js
jest.mock('@opentrons/components/src/deck/RobotWorkSpace', () => ({
  RobotWorkSpace: () => <></>
}))
```

Enzyme will still be able to match on RobotWorkSpace and get props passed in and such

mcous commented on app/src/components/CheckCalibration/utils.js:
```diff
@@ -0,0 +1,14 @@
+// @flow
+import type { JogAxis } from '../../http-api-client'
+
+const ORDERED_AXES: Array<JogAxis> = ['x', 'y', 'z']
```
Would this make sense as a tuple type instead of an array?

mcous commented on app/src/components/JogControls/index.js:
```diff
@@ -94,37 +112,57 @@ export class JogControls extends React.Component<
+    // because flow will bail out type to mixed with the latter
+    const handlersForAxes: Array<KeypressHandler> = axes
+      .map(a => handlersByAxis[a])
+      .flatMap(handler => handler)
```
Unless I'm missing something obvious (it's before noon and I'm only one coffee in, so forgive me if I am), could we just use a `flatMap`?

```js
    const handlersForAxes: Array<KeypressHandler> = axes
      .flatMap(a => handlersByAxis[a])
```

b-cooper replied:
Good point!

mcous commented on app/src/components/JogControls/index.js:
```diff
@@ -94,37 +112,57 @@ export class JogControls extends React.Component<
-        ))}
+        {axes
+          .map(a => JOG_BUTTON_NAMES_BY_AXIS[a])
+          .flatMap(name => name)
```
Same question here: this looks like it should be one flatMap call rather than a map and an identity flatMap

mcous commented on app/src/robot-api/operators.js:
```diff
@@ -12,7 +12,7 @@ import type { State, Action } from '../types'
   TriggerAction,
   State
-) => Types.RobotApiRequestOptions
+) => ?Types.RobotApiRequestOptions
```
Nitpick, but could we prefer explicit `null` to the Flow `?` type? Makes checks easier down the line (can use `=== null` instead of `== null` and `?Type` doesn't exist in TypeScript

mcous commented on app/src/robot-api/operators.js:
```diff
@@ -43,6 +43,7 @@ export function mapToRobotApiRequest<A>(
   return pipe(
     withRobotHost(state$, getRobotName),
     map(([a, s, host]) => [host, mapActionToRequest(a, s), a]),
+    filter(request => request != null),
```
I like the idea, but I don't think this is correct. Could you add a test to cover this behavior?

The `map` places the result of `mapActionToRequest` into a tuple, so `request` will never be `null`, it will be `[host, null, triggerAction]`

mcous reviewed (changes requested):
Haven't gotten to testing on my robot yet, but the code looks pretty good! A few nitpicks here and there, but I'm liking the test coverage.

I also like the bail out behavior addition to the API request mapper, but I don't think it'll work as written, so that's my change request for now. Will test on my robot and update with my results

mcous commented on app/src/calibration/epic/__tests__/fetchRobotCalibrationCheckSessionEpic.test.js:
```diff
@@ -0,0 +1,133 @@
+    }
+
+    it('calls GET /calibration/check/session', () => {
+      testScheduler.run(({ hot, cold, expectObservable, flush }) => {
```
You should use the API mock helpers for these tests. It'll bring the size of this PR down.

https://github.com/Opentrons/opentrons/blob/820eb6b76b676aceaa344107f972cf2538217984/app/src/robot-api/__utils__/epic-test-mocks.js#L45

b-cooper replied:
I'll do it for the 'GET' epic now, and the update after the api changes.

mcous commented on app/src/components/CheckCalibration/DeckSetup.js:
```diff
@@ -0,0 +1,121 @@
+import styles from './styles.css'
+
+const DECK_SETUP_PROMPT =
+  'Place full tiprack(s) on the deck, in their designated slots, as illustrated below.'
```
Screens are wrong on this one. Should be "tip rack" (two words) rather than one word for consistency across out stack in `DECK_SETUP_PROMPT` and `DECK_SETUP_BUTTON_TEXT`

b-cooper committed:
address a couple of pr comments

b-cooper committed:
use robot api epic test utils for fetch

b-cooper committed:
address mike pr comments

b-cooper committed:
old friend Mu

b-cooper commented:
@sfoster1 regarding your change requests:

 - "tip rack names are truncated on the deck" This has been a problem across our stack and because of the svg/html divide it's not quite simple to add another line. I totally agree this needs to be addressed, though it's a large bag of chips for this PR. I'm adding a comment and noting with design
- "You can still queue up jogs by pressing really fast". I did a quick spelunk and added a debounce, which works great, but breaks some of the epic tests. Instead of banging my head against the wall more, @mcous  and I have discussed a better way to address this UX concern that will require a couple of tweaks to the shared operators we're using across the app to communicate with the robot http api. Instead of debouncing for an arbitrary delay, this should either be handled by the domain layer not dispatching another HTTP request if one of the same type is already in flight, or this could alternatively be addressed on the robot side, by swallowing requests to jog while a jog is ongoing etc.  
- "If you fail to pick up tip and press try again the pipette doesn’t move back to the tip rack", That is a great point and I'm noting it, though that is managed by the python in this flow, so it will have to be addressed there.
- I fixed up the new instance of our old friend "capital Mu" in the tiprack names. A long term, less hacky solution, might be in the not too distant future.

b-cooper committed:
add todo

b-cooper committed:
explicit null return in default case

b-cooper committed:
remove old test cruft

sfoster1 reviewed (approved):
LGTM then!

b-cooper committed:
actually check request and not whole tuple

b-cooper committed:
1mm jog step instead of 2mm

mcous reviewed (approved):
🥮
